### PR TITLE
fix: auto-reconnect websocket/bot

### PR DESCRIPTION
**Pull request by TutorExilius**
## Pull request summary

Twitch bot did not relogged automatically. The tasks which are created by `asyncio.create_task()` need to be collected to prevent task disappearing, see: https://docs.python.org/3/library/asyncio-task.html#creating-tasks

>  Important: Save a reference to the tasks, to avoid a task disappearing mid-execution.

Fixed disappearing of tasks by collecting and deleteing them within a defined cycle.
With this fix, the websocket will reconnect automatically as intended.

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
    - [x] I have updated the changelog with a quick recap of my changes.
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by IAmTomahawkx** [comment]
Hello, I'll review this in a bit once I get home. However a few meta things:
1. Please run black over the affected files (make sure you use the -l120 flag) as it doesn't pass checks right now. 
2. Please update the changelog with a brief summary of changes made.

**Comment by TutorExilius** [comment]
Adapted changelog + blacked code (websocket.py)

File: twitchio/websocket.py [hunk 1]
```diff
@@ -106,6 +106,19 @@ def __init__(
...
+        self._background_tasks: List[asyncio.Task] = []
+        self._task_cleaner: Optional[asyncio.Task] = None
+
+    async def _task_cleanup(self):
+        while self.is_ready:
+            # keep all undone tasks
+            self._background_tasks = list(filter(lambda task: not task.done(), self._background_tasks))
+
+            # cleanup tasks every 5 minute
+            await asyncio.sleep(5 * 60)
```
Comment (IAmTomahawkx) [inline twitchio/websocket.py:120]: 5 minutes is probably too long, every 30 seconds or so would be better

File: twitchio/websocket.py [hunk 2]
```diff
@@ -119,6 +132,8 @@ async def _connect(self):
 
         if self._keeper:
             self._keeper.cancel()  # Stop our current keep alive.
+        if self._task_cleaner:
+            self._task_cleaner.cancel()  # Stop our current task cleaner.
```
Comment (IAmTomahawkx) [inline twitchio/websocket.py:136]: this doesn't really need to ever be cancelled except for at shutdown

File: twitchio/websocket.py [hunk 3]
```diff
@@ -141,11 +156,12 @@ async def _connect(self):
             log.error(f"Websocket connection failure: {e}:: Attempting reconnect in {retry} seconds.")
 
             await asyncio.sleep(retry)
-            return asyncio.create_task(self._connect())
+            await self._connect()
```
Comment (IAmTomahawkx) [inline twitchio/websocket.py:159]:
```suggestion
            return await self._connect()
```
you've removed the return here, which will continue the login process as though the connection was successful

File: twitchio/websocket.py [hunk 4]
```diff
@@ -171,8 +187,9 @@ async def _keep_alive(self):
                         continue
                     task = asyncio.create_task(self._process_data(event))
                     task.add_done_callback(partial(self._task_callback, event))  # Process our raw data
+                    self._background_tasks.append(task)
 
-        asyncio.create_task(self._connect())
+        await self._connect()
```
Comment (IAmTomahawkx) [inline twitchio/websocket.py:192]: why?

File: twitchio/websocket.py [hunk 5]
```diff
@@ -283,18 +302,21 @@ async def join_channels(self, *channels: str):
...
+                    task = asyncio.create_task(self._join_channel(channel, 11))
+                    self._background_tasks.append(task)
 
     async def _join_channel(self, entry: str, timeout: int):
         channel = re.sub("[#]", "", entry).lower()
         await self.send(f"JOIN #{channel}\r\n")
 
         self._join_pending[channel] = fut = self._loop.create_future()
-        asyncio.create_task(self._join_future_handle(fut, channel, timeout))
+        await self._join_future_handle(fut, channel, timeout)
```
Comment (IAmTomahawkx) [inline twitchio/websocket.py:319]: This should be a background task

File: twitchio/websocket.py [hunk 6]
```diff
@@ -516,6 +538,11 @@ def _fetch_futures(self):
 
     async def _close(self):
         self._keeper.cancel()
+        self._task_cleaner.cancel()
+
+        for task in self._background_tasks:
+            task.cancel()
```
Comment (IAmTomahawkx) [inline twitchio/websocket.py:544]:
```suggestion
        for task in self._background_tasks:
            if not task.done():
                task.cancel()
```

Comment (IAmTomahawkx) [inline twitchio/websocket.py:544]: cancelling the tasks probably isn't the best strategy, as that will dangle the error handlers if they get cancelled.

**Review by IAmTomahawkx — CHANGES_REQUESTED** [review]

File: twitchio/websocket.py (hunk 2 above)
Comment (TutorExilius) [reply twitchio/websocket.py:136]: That's right, I will adapt that part.

File: twitchio/websocket.py (hunk 4 above)
Comment (TutorExilius) [reply twitchio/websocket.py:192]: Don't we want to make sure here that a (re)connect should definitely take place at this point?
Furthermore, we spare ourselves an attribute to hold a reference to the task.

File: twitchio/websocket.py (hunk 6 above)
Comment (TutorExilius) [reply twitchio/websocket.py:544]: Good point

File: twitchio/websocket.py (hunk 4 above)
Comment (TutorExilius) [reply twitchio/websocket.py:193]: Don't we want to make sure here that a (re)connect should definitely take place at this point?
Furthermore, we spare ourselves an attribute to hold a reference to the task.

Comment (TutorExilius) [reply twitchio/websocket.py:193]: @IAmTomahawkx what do you think?

**Comment by IAmTomahawkx** [comment]
youll need to revert [ec0d9b1](https://github.com/TwitchIO/TwitchIO/pull/364/commits/ec0d9b1f4500ddc6a929178e94538ff5d7a254dd) and [1753a3b](https://github.com/TwitchIO/TwitchIO/pull/364/commits/1753a3bf3d3ea7a0f5f972884bfc64df804006ad) and rerun black with the -l120 flag, as running black on its own first has some unintended consequences

File: twitchio/websocket.py [hunk 7]
```diff
@@ -181,7 +199,7 @@ def _task_callback(self, data, task):
             log.error("Authentication error. Please check your credentials and try again.")
             self._close()
         elif exc:
-            asyncio.create_task(self.event_error(exc, data))
+            self._background_tasks.append(asyncio.create_task(self.event_error(exc, data)))
```
Comment (IAmTomahawkx) [inline twitchio/websocket.py:202]: Due to the nature of your cleaning up cancelling all the tasks, you'll need to wrap the event_error in asyncio.shield

File: twitchio/websocket.py (hunk 4 above)
Comment (IAmTomahawkx) [reply twitchio/websocket.py:193]: creating the task will cause a reconnect to take place when the event loop is ready. The keep alive function shouldn't be responsible for the lifespan of the connect call, as the connect function will cancel the keep alive, causing it to cancel itself at the next context switch

**Review by IAmTomahawkx — CHANGES_REQUESTED** [review]

File: twitchio/websocket.py (hunk 4 above)
Comment (TutorExilius) [reply twitchio/websocket.py]: Thanks for explaning, done

File: twitchio/websocket.py (hunk 7 above)
Comment (TutorExilius) [reply twitchio/websocket.py]: done

**Comment by TutorExilius** [comment]
Squasched two commits together ([ec0d9b1](https://github.com/TwitchIO/TwitchIO/pull/364/commits/ec0d9b1f4500ddc6a929178e94538ff5d7a254dd) and [1753a3b](https://github.com/TwitchIO/TwitchIO/pull/364/commits/1753a3bf3d3ea7a0f5f972884bfc64df804006ad) blacked code) and force pushed.

**Review by IAmTomahawkx — APPROVED** [review]

**Comment by TutorExilius** [comment]
what is the status? when can we expect the fix? :)

**Comment by chillymosh** [comment]
LGTM but will wait for @EvieePy to approve this one

**Review by EvieePy — APPROVED** [review]
Looks fine to me, though I won't commit this until we have done a few tests locally.

**Pull request merged**